code-generator: rewrite hack/update-codecs.sh into reusable generate-{internal,}-groups.sh

sttts opened the pull request:
Generating everything for groups inside of an apiserver (with internal types) becomes:

```shell
generate-internal-groups.sh all "$(dirname ${BASH_SOURCE})/../../.." k8s.io/sample-apiserver/pkg/client k8s.io/sample-apiserver/pkg/apis k8s.io/sample-apiserver/pkg/apis wardle:v1alpha1
```

Generating everything for a CRD (versioned types) becomes:

```shell
generate-groups.sh all "$(dirname ${BASH_SOURCE})/../../.." k8s.io/sample-apiserver/pkg/client k8s.io/sample-apiserver/pkg/apis wardle:v1alpha1
```

This should cover the 90% percent use-case. For the other 10% this can be forked and adapted as needed.

Furthermore, we can put this into a Docker container. Then code-generator consumers can then do:

```shell
$ docker run -v $GOPATH:/go k8s.io/code-generator:1.8 generate-group.sh github.com/foo/bar example:v1
```

This is possibly only the first step towards a `code-generator` binary. For the later deeper generator changes are necessary (e.g. https://github.com/kubernetes/kubernetes/pull/53202) and hence the later is only feasible in 1.9. **This PR here in contrast, we can cherry-pick to 1.8.**

```release-note
Add generate-groups.sh and generate-internal-groups.sh to k8s.io/code-generator to easily run generators against CRD or User API Server types.
```

Fixes https://github.com/kubernetes/kubernetes/issues/48714.

sttts commented:
/cc @mrIncompetent

mrIncompetent commented:
This is wonderful! Very much needed!

deads2k commented:
@eparis can you help review the bash?

sttts commented:
/cc @colemickens @nikhita – could you take a look at this? It is especially meant to make the life of CRD users easier.

sttts commented:
/cc @ericchiang @kubernetes/sig-apps-misc – review and feedback is welcome

munnerz commented on staging/src/k8s.io/code-generator/hack/lib/codegen.sh:
```diff
@@ -0,0 +1,117 @@
+    IFS=: read G Vs <<<"${GVs}"
+
+    # enumerate versions
+    for V in ${Vs//,/ } do
```
I think there's a missing `;` after `for V in ${Vs//,/ }` here. I get the following when running on OS X:

Adding it resolves the following:

`/Users/james/go/src/k8s.io/kubernetes/vendor/k8s.io/code-generator/hack/lib/codegen.sh:98: parse error near 'done'`

sttts replied:
indeed, missing ;

hongchaodeng commented on staging/src/k8s.io/code-generator/hack/lib/codegen.sh:
```diff
@@ -0,0 +1,117 @@
+
+for gen in defaulter-gen conversion-gen client-gen lister-gen informer-gen deepcopy-gen conversion-gen; do
+  echo "Building ${gen}"
+  go install ${CODEGEN_PKG}/cmd/${gen}
```
This incurs side effect for source-ing this lib.

sttts replied:
true. We should do this only on-demand and only once.

munnerz commented:
It'd be great if this library could also expose the `--verify` flag of it's underlying generators. I've found this really useful in CI, to ensure that generated files are up to date in the repo.

hongchaodeng commented on staging/src/k8s.io/code-generator/hack/lib/codegen.sh:
```diff
@@ -0,0 +1,117 @@
+}
+
+function generate-groups() {
+  local BASE="$1" # the base packages are resolved against (empty or the vendor/ directory ${OUTPUT_PKG} is in)
```
I have to confess these vars are difficult to understand. Can you give some specific examples?

sttts replied:
simplified this. BASE is usually not needed, only in the kube case where we have staging/ directories.

hongchaodeng reviewed (commented):
@sttts 
Awesome work!
A shared docker container like `k8s.io/code-generator:1.8` would be very useful too!

hongchaodeng commented:
@sttts
Can you assign to me? I can help test the script out.

hongchaodeng commented on staging/src/k8s.io/code-generator/hack/lib/codegen.sh:
```diff
@@ -0,0 +1,117 @@
+  done
+
+  echo "Generating deepcopy funcs"
+  ${GOPATH}/bin/deepcopy-gen -i $(join , "${ALL_FQ_APIS[@]}") -O zz_generated.deepcopy --bounding-dirs ${INT_APIS_PKG},${EXT_APIS_PKG} "$@"
```
Where is boilerplate flag defined?

sttts replied:
Added a demonstration to sample-apiserver

sttts committed:
code-generator: rename _test to _examples

sttts committed:
code-generator: turn hack/update-codegen.sh into re-usable generate-{internal,}-groups.sh scripts

sttts committed:
sample-apiserver: port to k8s.io/code-generator/generate-internal-groups.sh

sttts commented:
@munnerz @hongchaodeng addressed all comments. ptal.

sttts committed:
verify-pkg-names.sh: exclude generated informers

hongchaodeng commented on staging/src/k8s.io/code-generator/hack/lib/codegen.sh:
```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
```
```
#!/usr/bin/env bash
```

hongchaodeng commented on staging/src/k8s.io/code-generator/hack/lib/codegen.sh:
```diff
@@ -0,0 +1,147 @@
+set -o nounset
+set -o pipefail
+
+CODEGEN_PKG=./$(dirname "${BASH_SOURCE}")/../..
```
```
(dirname "${BASH_SOURCE}")
```

This won't work on other shell like zsh. For me, the terminal exited.

hongchaodeng replied:
Use `$0` instead?

hongchaodeng replied:
Actually this var is only used in `codegen::ensure-generators-built`.
Why make it global exported VAR? Maybe make it local there?

hongchaodeng commented:
I confirm that after fixing my above minor comments it works for me 👍  :
```
codegen::generate-groups all \
  github.com/coreos/etcd-operator/pkg/client \
  github.com/coreos/etcd-operator/pkg/apis/ etcd:v1beta2
```

munnerz commented:
I can also confirm it's working nicely in https://github.com/kubernetes/kubernetes/pull/52753 😄

hongchaodeng commented:
@sttts
The tests are failing.

hongchaodeng commented on staging/src/k8s.io/code-generator/hack/lib/codegen.sh:
```diff
@@ -0,0 +1,147 @@
+CODEGEN_PKG=./$(dirname "${BASH_SOURCE}")/../..
+
+# build generators on-demand
+CODEGEN_BUILT=false
```
Also is it possible not to keep this state in this library?
Couldn't user just call install explicitly outside.

sttts committed:
Update generated files

sttts commented:
@ericchiang @munnerz addressed your comments. Moreover, I have switch from the codegen.sh library approach to two executable scripts in `k8s.io/code-generator`.  This is more in-line with the future `code-generator` binary, if we decide to rewrite the shell scripts to Go.

From my side this is complete now to go into master and the be cherry-picked into 1.8. A possible Golang rewrite has to wait for 1.9.

hongchaodeng commented:
lgtm if tests pass

sttts commented:
/retest

sttts commented:
@pwittrock after seeing your community meeting slides with `Support for bootstrapping CRD go clients, go controllers, ref docs?`, would be great to get your input on these kind of changes.
